In [1]:
#load required libraries
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
os.getcwd()

'C:\\Users\\plw40'

In [3]:
os.chdir("C:\\Users\plw40\\OneDrive\\Documents\\Magnimind\\datasets")

In [4]:
df = pd.read_csv("ksprojects.csv")
df.head(5)

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [5]:
#how many rows of data are in the dataframe?
print(df.shape)

(65418, 13)


There are 65,418 rows in this dataset

In [6]:
#what are the names and data types of the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


In [7]:
#do any of the columns contain null values?
df.isnull().sum()

id                 0
name               0
category           0
main_category      0
deadline           0
goal               0
launched           0
state              0
backers            0
country            0
usd_pledged      745
currency           0
pledged            0
dtype: int64

usd_pledged has 745 null values 
- would want to understand if they are missing values or if the value represents 0 pledged.
- if the null is actually representative of $0 pledge then we can replace the null with 0.
- if this is missing data we would need to determine if we should drop these rows or try to impute the value.
- possibly null because the pledge was not made in usd. Convert to USD to impute.

In [8]:
# find all successful documentary projects and sort them by the amount pledged. Print top 10 highest pledges.
#first make copy to maintain integrity of original
df_copy = df.copy()

In [9]:
succ_doc = df_copy.loc[(df_copy['category']=="Documentary") & (df_copy['state']=="successful")]    

In [10]:
succ_doc['pledged'].max()

384376.0

In [11]:
succ_doc.shape

(1087, 13)

In [12]:
sorted_docs = succ_doc.sort_values('pledged')
sorted_docs.tail(10)


id                                               name  \
22237   122290911                          The Killing of Tony Blair   
43820  1194449879  PlantPure Nation, a Documentary with Dr. T. Co...   
27268  1228065368                  El Voto Hispano | The Documentary   
16243  1061853312                               Linsanity: the Movie   
27402  1135999291                             Rewind To Fast-Forward   
28003  1205913248                                     ILEANA'S SMILE   
61248  1203470139                         Curious George Documentary   
12127  1326363492                                   The Culture High   
24166  1162836718  "The Wrecking Crew" The Untold Story of Rock &...   
11731  1078982332                BRIDEGROOM - An American Love Story   

          category main_category             deadline      goal  \
22237  Documentary  Film & Video  2013-10-07 18:17:14   50000.0   
43820  Documentary  Film & Video  2015-05-14 17:41:41  150000.0   
27268  Documentary  Film & Video  2014-04-27 08:15:23  158000.0   
16243  Documentary  Film & Video  2013-02-16 00:00:00  117000.0   
27402  Documentary  Film & Video  2014-05-01 07:59:00  137000.0   
28003  Documentary  Film & Video  2014-05-21 09:05:00  133000.0   
61248  Documentary  Film & Video  2016-08-25 00:00:00  175000.0   
12127  Documentary  Film & Video  2012-08-04 04:00:00  190000.0   
24166  Documentary  Film & Video  2013-12-21 08:59:00  250000.0   
11731  Documentary  Film & Video  2012-07-20 01:00:28  300000.0   

                  launched       state  backers country  usd_pledged currency  \
22237  2013-08-28 18:17:14  successful     4459      GB     255150.0      GBP   
43820  2015-03-15 16:41:41  successful     1099      US     163930.0      USD   
27268  2014-03-18 07:15:23  successful      461      US     165401.0      USD   
16243  2013-01-15 19:46:34  successful     2102      US     167917.0      USD   
27402  2014-03-30 13:35:31  successful     4395      US     176746.0      USD   
28003  2014-04-21 09:05:00  successful     1173      US     177421.0      USD   
61248  2016-07-26 18:01:41  successful     1483      US       2028.0      USD   
12127  2012-06-22 01:40:20  successful     3448      US     240022.0      USD   
24166  2013-10-30 21:06:34  successful     4245      US     313158.0      USD   
11731  2012-06-20 01:00:28  successful     6508      US     384376.0      USD   

        pledged  
22237  163891.0  
43820  163930.0  
27268  165401.0  
16243  167917.0  
27402  176746.0  
28003  177421.0  
61248  186010.0  
12127  240022.0  
24166  313158.0  
11731  384376.0

In [13]:
#create a new col named average_per_backer and set its value to the total amount pledged/number of backers
df_copy['average_per_backer'] = round(df_copy['pledged']/df_copy['backers'],2)
df_copy['average_per_backer'].head(10)

0     10.00
1     50.00
2     42.72
3    273.96
4     22.80
5     70.00
6     15.00
7     46.25
8      9.33
9     45.35
Name: average_per_backer, dtype: float64

In [14]:
df_copy['average_per_backer'].isnull().sum()

9282

In [15]:
print(df_copy['average_per_backer'].head(30))

0      10.00
1      50.00
2      42.72
3     273.96
4      22.80
5      70.00
6      15.00
7      46.25
8       9.33
9      45.35
10     49.43
11     30.00
12     13.67
13     20.00
14     50.00
15      7.00
16     52.75
17    158.89
18       NaN
19       NaN
20    961.00
21     44.74
22     40.79
23     43.12
24     83.95
25     18.58
26     32.14
27     16.25
28     20.29
29     43.89
Name: average_per_backer, dtype: float64


**What happened to the rows with 0 backers? how can this be dealt with?**
Returns a null because you can't divide by 0 could us an if statement or replace all NaN with 0.

In [16]:
#drop all rows with 0 backers then repeat the previous exercise
with_backers = df.copy()
with_backers = with_backers[with_backers.backers != 0]
print(with_backers['backers'].min())

1


In [17]:
with_backers.dropna()

id                                               name  \
0      1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1      1100844465                                       daily digest   
2      1316334968                             Drive A Faster Car 2.0   
3      1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4      1099226462          Logical Guess Pictures' 2nd Horror Movie!   
...           ...                                                ...   
65408  1166775907        The sports wall racks collection by MEOLLO™   
65410  1120449668  Not For You: Family Narratives of Denial & Com...   
65411  1105965008      Anyone can Learn English. Anywhere. For Free.   
65413  1107492255    DODEKA - music finally within everyone's reach!   
65414  1373187022  World's Smallest Water-resistant Magnetic Came...   

             category main_category             deadline     goal  \
0         Documentary  Film & Video  2009-05-20 21:50:00    300.0   
1         Documentary  Film & Video  2009-06-01 02:20:00    700.0   
2          Nonfiction    Publishing  2009-06-04 06:00:00   1000.0   
3             Theater       Theater  2009-06-05 05:59:00   6000.0   
4        Film & Video  Film & Video  2009-06-06 00:45:00    500.0   
...               ...           ...                  ...      ...   
65408  Product Design        Design  2017-02-01 23:59:00   8000.0   
65410            Food          Food  2017-02-03 07:24:36   3500.0   
65411      Technology    Technology  2017-02-03 07:40:46  16000.0   
65413           Sound    Technology  2017-02-03 14:51:00  55000.0   
65414         Gadgets    Technology  2017-02-03 17:23:11   2000.0   

                  launched       state  backers country  usd_pledged currency  \
0      2009-04-30 22:10:30      failed        4      US       40.000      USD   
1      2009-05-04 21:14:28  successful       14      US      700.000      USD   
2      2009-05-04 22:51:31  successful       32      US     1367.000      USD   
3      2009-04-30 22:22:43  successful       24      US     6575.000      USD   
4      2009-04-30 01:32:55  successful       22      US      501.660      USD   
...                    ...         ...      ...     ...          ...      ...   
65408  2016-12-05 18:00:10        live        4      ES      153.600      EUR   
65410  2016-12-05 07:24:36        live       24      US      235.000      USD   
65411  2016-12-05 07:40:46        live       10      GB      159.122      GBP   
65413  2016-12-05 14:51:00        live        9      CH     1022.120      CHF   
65414  2016-12-05 17:23:11    canceled      148      US     5474.000      USD   

       pledged  
0        40.00  
1       700.00  
2      1367.00  
3      6575.00  
4       501.66  
...        ...  
65408   144.00  
65410   527.00  
65411   475.00  
65413  9287.00  
65414  5552.00  

[55531 rows x 13 columns]

In [18]:
with_backers['average_per_backer'] = round(with_backers['pledged']/with_backers['backers'],2)
print(with_backers['average_per_backer'].head(10))

0     10.00
1     50.00
2     42.72
3    273.96
4     22.80
5     70.00
6     15.00
7     46.25
8      9.33
9     45.35
Name: average_per_backer, dtype: float64


In [19]:
with_backers['average_per_backer'].isnull().sum()

0

**Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.**


In [20]:
index = ['a','b','c','d','e','f','g','h','i','j']
data = {'prime':[2,3,5,7,11,13,17,19,23,29],'fibonacci': [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]}
new_df = pd.DataFrame(data, index=index)
new_df

prime  fibonacci
a      2          0
b      3          1
c      5          1
d      7          2
e     11          3
f     13          5
g     17          8
h     19         13
i     23         21
j     29         34

In [21]:
#merge the two dfs given into one called nycflights13

In [22]:
flights = pd.read_csv("flights.csv")
print(flights.head(5))

   year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416 

In [23]:
flights.shape

(336776, 19)

In [24]:
airlines = pd.read_csv("airlines.csv")
print(airlines.head(5))
print(airlines.shape)

  carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.
2      AS    Alaska Airlines Inc.
3      B6         JetBlue Airways
4      DL    Delta Air Lines Inc.
(16, 2)


The first thing to notice is there are far less observations in airlines than flights. So, we want to merge on col carrier duplicating the airline name according to the carrier code.

In [25]:
df_merged = flights.merge(airlines[['carrier', 'name']], on = 'carrier', how = 'left')
print(df_merged.head(5))

   year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour                    name  
0      1400     5      15  2013-01-01

In [26]:
#are there any missing values in this df if so drop them
df_merged.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
name                 0
dtype: int64

In [27]:
df_merged_nona = df_merged.dropna()

In [28]:
df_merged_nona.isnull().sum()

year              0
month             0
day               0
dep_time          0
sched_dep_time    0
dep_delay         0
arr_time          0
sched_arr_time    0
arr_delay         0
carrier           0
flight            0
tailnum           0
origin            0
dest              0
air_time          0
distance          0
hour              0
minute            0
time_hour         0
name              0
dtype: int64

In [29]:
#merge these data sets into one called 'census'
state_pop = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/state-populations.csv")
census_div = pd.read_csv("https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-science/census-divisions.csv")
print(state_pop.head(5))
print(census_div.head(5))

        state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016  
0   4863300  
1    741894  
2   6931071  
3   2988248  
4  39250017  
           state     region     division
0    Connecticut  Northeast  New England
1          Maine  Northeast  New England
2  Massachusetts  Northeast  New England
3  New Hampshire  Northeast  New England
4   Rhode Island  Northeast  New England


In [30]:
state_pop.shape

(51, 8)

In [31]:
census_div.shape

(51, 3)

In [32]:
census = state_pop.merge(census_div[['state','region','division']])
census.head(5)

state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016 region            division  
0   4863300  South  East South Central  
1    741894   West             Pacific  
2   6931071   West            Mountain  
3   2988248  South  West South Central  
4  39250017   West             Pacific

In [51]:
#Re-shape census so that one column contains all population measures, and another the year attributes
census_copy = census.copy()
census_copy.head(10)

state      2010      2011      2012      2013      2014  \
0               Alabama   4785492   4799918   4815960   4829479   4843214   
1                Alaska    714031    722713    731089    736879    736705   
2               Arizona   6408312   6467163   6549634   6624617   6719993   
3              Arkansas   2921995   2939493   2950685   2958663   2966912   
4            California  37332685  37676861  38011074  38335203  38680810   
5              Colorado   5048644   5118360   5189867   5267603   5349648   
6           Connecticut   3579899   3589893   3593795   3596003   3591873   
7              Delaware    899816    907924    916993    925395    934948   
8  District of Columbia    605183    620477    635327    649165    659005   
9               Florida  18849098  19096952  19344156  19582022  19888741   

       2015      2016     region            division  
0   4853875   4863300      South  East South Central  
1    737709    741894       West             Pacific  
2   6817565   6931071       West            Mountain  
3   2977853   2988248      South  West South Central  
4  38993940  39250017       West             Pacific  
5   5448819   5540545       West            Mountain  
6   3584730   3576452  Northeast         New England  
7    944076    952065      South      South Atlantic  
8    670377    681170      South      South Atlantic  
9  20244914  20612439      South      South Atlantic

In [34]:
cen_transpose = census_copy.transpose()
cen_transpose.head(5)

0       1        2         3           4         5            6   \
state  Alabama  Alaska  Arizona  Arkansas  California  Colorado  Connecticut   
2010   4785492  714031  6408312   2921995    37332685   5048644      3579899   
2011   4799918  722713  6467163   2939493    37676861   5118360      3589893   
2012   4815960  731089  6549634   2950685    38011074   5189867      3593795   
2013   4829479  736879  6624617   2958663    38335203   5267603      3596003   

             7                     8         9   ...            41         42  \
state  Delaware  District of Columbia   Florida  ...  South Dakota  Tennessee   
2010     899816                605183  18849098  ...        816325    6356671   
2011     907924                620477  19096952  ...        824398    6397634   
2012     916993                635327  19344156  ...        834441    6454306   
2013     925395                649165  19582022  ...        844922    6494821   

             43       44       45        46          47             48  \
state     Texas     Utah  Vermont  Virginia  Washington  West Virginia   
2010   25244310  2775326   625982   8025773     6743226        1854230   
2011   25646389  2816124   626730   8110035     6822520        1854972   
2012   26071655  2855782   626444   8192048     6895226        1856560   
2013   26473525  2902663   627140   8262692     6968006        1853231   

              49       50  
state  Wisconsin  Wyoming  
2010     5690263   564513  
2011     5709640   567725  
2012     5726177   576765  
2013     5742854   582684  

[5 rows x 51 columns]

In [35]:
census_copy.head(5)

state      2010      2011      2012      2013      2014      2015  \
0     Alabama   4785492   4799918   4815960   4829479   4843214   4853875   
1      Alaska    714031    722713    731089    736879    736705    737709   
2     Arizona   6408312   6467163   6549634   6624617   6719993   6817565   
3    Arkansas   2921995   2939493   2950685   2958663   2966912   2977853   
4  California  37332685  37676861  38011074  38335203  38680810  38993940   

       2016 region            division  
0   4863300  South  East South Central  
1    741894   West             Pacific  
2   6931071   West            Mountain  
3   2988248  South  West South Central  
4  39250017   West             Pacific

In [36]:
census_copy[['2010','2011','2012','2013','2014','2015','2016']].head(5)

2010      2011      2012      2013      2014      2015      2016
0   4785492   4799918   4815960   4829479   4843214   4853875   4863300
1    714031    722713    731089    736879    736705    737709    741894
2   6408312   6467163   6549634   6624617   6719993   6817565   6931071
3   2921995   2939493   2950685   2958663   2966912   2977853   2988248
4  37332685  37676861  38011074  38335203  38680810  38993940  39250017

In [37]:
census_yrs = census_copy[['2010','2011','2012','2013','2014','2015','2016']]

In [38]:
census_yrs = census_yrs.stack()
print(census_yrs.head(10))

0  2010    4785492
   2011    4799918
   2012    4815960
   2013    4829479
   2014    4843214
   2015    4853875
   2016    4863300
1  2010     714031
   2011     722713
   2012     731089
dtype: int64


In [39]:
#group data by year and summarize it
census_yrs.info()

<class 'pandas.core.series.Series'>
MultiIndex: 357 entries, (0, '2010') to (50, '2016')
Series name: None
Non-Null Count  Dtype
--------------  -----
357 non-null    int64
dtypes: int64(1)
memory usage: 6.3+ KB


In [40]:
years_df = pd.DataFrame(census_yrs)

In [41]:
years_df.head(5)

0
0 2010  4785492
  2011  4799918
  2012  4815960
  2013  4829479
  2014  4843214

In [42]:
years_df_reindex = years_df.reset_index(level = [1])
years_df_reindex.columns=['years','population']
years_df_reindex.head(5)


years  population
0  2010     4785492
0  2011     4799918
0  2012     4815960
0  2013     4829479
0  2014     4843214

In [43]:
popbyyear = years_df_reindex.groupby('years').sum()

In [44]:
popbyyear

population
years            
2010    309348193
2011    311663358
2012    313998379
2013    316204908
2014    318563456
2015    320896618
2016    323127513

In [45]:
popbyyear.describe() #unsure if this is right?

population
count  7.000000e+00
mean   3.162575e+08
std    4.966336e+06
min    3.093482e+08
25%    3.128309e+08
50%    3.162049e+08
75%    3.197300e+08
max    3.231275e+08

In [46]:
#group data by region, division, and year and summarize it
census_copy.columns

Index(['state', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
       'region', 'division'],
      dtype='object')

In [47]:
groupby = ['2010', '2011', '2012', '2013', '2014', '2015', '2016',
       'region', 'division']
census_copy.groupby(groupby)
census_copy.index

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
            50],
           dtype='int64')

In [48]:
census_copy.describe()

2010          2011          2012          2013          2014  \
count  5.100000e+01  5.100000e+01  5.100000e+01  5.100000e+01  5.100000e+01   
mean   6.065651e+06  6.111046e+06  6.156831e+06  6.200096e+06  6.246342e+06   
std    6.839536e+06  6.907730e+06  6.974942e+06  7.038774e+06  7.110811e+06   
min    5.645130e+05  5.677250e+05  5.767650e+05  5.826840e+05  5.836420e+05   
25%    1.700530e+06  1.713213e+06  1.725818e+06  1.732621e+06  1.741023e+06   
50%    4.348662e+06  4.369354e+06  4.384799e+06  4.400477e+06  4.413057e+06   
75%    6.654375e+06  6.717222e+06  6.776617e+06  6.837396e+06  6.902054e+06   
max    3.733268e+07  3.767686e+07  3.801107e+07  3.833520e+07  3.868081e+07   

               2015          2016  
count  5.100000e+01  5.100000e+01  
mean   6.292091e+06  6.335834e+06  
std    7.181603e+06  7.243849e+06  
min    5.865550e+05  5.855010e+05  
25%    1.746940e+06  1.757121e+06  
50%    4.424611e+06  4.436974e+06  
75%    6.988928e+06  7.109536e+06  
max    3.899394e+07  3.925002e+07

In [49]:
ksprojects = pd.read_csv("ksprojects.csv")
ksprojects.head(5)

id                                               name  \
0  1147015301  "All We Had" Gets Into Cannes -- $10 or More G...   
1  1100844465                                       daily digest   
2  1316334968                             Drive A Faster Car 2.0   
3  1304906577         Accidental to Edinburgh - PHASE 1: AIRFARE   
4  1099226462          Logical Guess Pictures' 2nd Horror Movie!   

       category main_category             deadline    goal  \
0   Documentary  Film & Video  2009-05-20 21:50:00   300.0   
1   Documentary  Film & Video  2009-06-01 02:20:00   700.0   
2    Nonfiction    Publishing  2009-06-04 06:00:00  1000.0   
3       Theater       Theater  2009-06-05 05:59:00  6000.0   
4  Film & Video  Film & Video  2009-06-06 00:45:00   500.0   

              launched       state  backers country  usd_pledged currency  \
0  2009-04-30 22:10:30      failed        4      US        40.00      USD   
1  2009-05-04 21:14:28  successful       14      US       700.00      USD   
2  2009-05-04 22:51:31  successful       32      US      1367.00      USD   
3  2009-04-30 22:22:43  successful       24      US      6575.00      USD   
4  2009-04-30 01:32:55  successful       22      US       501.66      USD   

   pledged  
0    40.00  
1   700.00  
2  1367.00  
3  6575.00  
4   501.66

In [50]:
#create a crosstab between state and category showing amount pledged as the value
pd.crosstab(index=ksprojects['state'], columns=ksprojects['category'], values=ksprojects['pledged'], aggfunc=sum)


category    3D Printing   Academic  Accessories    Action   Animals  \
state                                                                 
canceled      106827.50    4364.00    124936.11   13342.0      0.00   
failed        371417.00   81427.74    320990.20   79978.1  15769.87   
live          309215.00      31.00     60904.00       0.0       NaN   
successful   8511414.01  354485.24   2569489.70  226333.9  75497.10   
suspended      10146.00     226.00      3822.00       NaN      0.00   
undefined      13757.00        NaN      2763.00       NaN       NaN   

category     Animation  Anthologies     Apparel       Apps  Architecture  ...  \
state                                                                     ...   
canceled      58910.00      4583.00   228836.00  174745.49      12311.00  ...   
failed       498129.57     20991.00   401102.55  553596.12     139656.99  ...   
live          98403.80      3339.00   236921.59   60101.00       2117.00  ...   
successful  2605326.54   1346909.64  6446055.31  755165.60     460293.08  ...   
suspended         0.00          NaN    56213.00   12989.00        131.00  ...   
undefined          NaN          NaN         NaN        NaN           NaN  ...   

category    Wearables  Weaving        Web   Webcomics   Webseries  \
state                                                               
canceled     239056.0    212.0   88658.00     5668.00    54679.39   
failed       718671.5   6686.0  382408.65    26085.00  1065147.12   
live         141310.0      NaN   57931.00    82120.00      185.00   
successful  7381979.5   6019.0  912191.70  1164701.39  3821840.97   
suspended      9344.0      0.0      50.00         NaN    14878.00   
undefined         NaN      NaN        NaN         NaN         NaN   

category    Woodworking  Workshops  World Music  Young Adult     Zines  
state                                                                   
canceled         153.01        NaN     17136.00       661.00   6940.00  
failed         60328.00     2360.0    120691.83     20920.20  19692.00  
live            1211.73        0.0      9580.00       575.00     80.00  
successful    214568.35    21885.0   1069703.20    145305.62  48668.53  
suspended         91.00        0.0        37.00          NaN       NaN  
undefined           NaN        NaN          NaN          NaN       NaN  

[6 rows x 158 columns]

In [53]:
census_copy.head(10)

state      2010      2011      2012      2013      2014  \
0               Alabama   4785492   4799918   4815960   4829479   4843214   
1                Alaska    714031    722713    731089    736879    736705   
2               Arizona   6408312   6467163   6549634   6624617   6719993   
3              Arkansas   2921995   2939493   2950685   2958663   2966912   
4            California  37332685  37676861  38011074  38335203  38680810   
5              Colorado   5048644   5118360   5189867   5267603   5349648   
6           Connecticut   3579899   3589893   3593795   3596003   3591873   
7              Delaware    899816    907924    916993    925395    934948   
8  District of Columbia    605183    620477    635327    649165    659005   
9               Florida  18849098  19096952  19344156  19582022  19888741   

       2015      2016     region            division  
0   4853875   4863300      South  East South Central  
1    737709    741894       West             Pacific  
2   6817565   6931071       West            Mountain  
3   2977853   2988248      South  West South Central  
4  38993940  39250017       West             Pacific  
5   5448819   5540545       West            Mountain  
6   3584730   3576452  Northeast         New England  
7    944076    952065      South      South Atlantic  
8    670377    681170      South      South Atlantic  
9  20244914  20612439      South      South Atlantic

In [55]:
#Re-shape census so that one column contains all population measures, and another the year attributes
reshaped_census = pd.melt(census_copy, 
                          id_vars=['state', 'region', 'division'], 
                          value_vars=['2010', '2011', '2012', '2013', '2014', '2015', '2016'], 
                          var_name='year', 
                          value_name='population')
reshaped_census.head(10)

state     region            division  year  population
0               Alabama      South  East South Central  2010     4785492
1                Alaska       West             Pacific  2010      714031
2               Arizona       West            Mountain  2010     6408312
3              Arkansas      South  West South Central  2010     2921995
4            California       West             Pacific  2010    37332685
5              Colorado       West            Mountain  2010     5048644
6           Connecticut  Northeast         New England  2010     3579899
7              Delaware      South      South Atlantic  2010      899816
8  District of Columbia      South      South Atlantic  2010      605183
9               Florida      South      South Atlantic  2010    18849098

In [56]:
#group data by year and summarize it
reshaped_census.groupby('year').describe()


population                                                              \
          count          mean           std       min        25%        50%   
year                                                                          
2010       51.0  6.065651e+06  6.839536e+06  564513.0  1700530.5  4348662.0   
2011       51.0  6.111046e+06  6.907730e+06  567725.0  1713213.0  4369354.0   
2012       51.0  6.156831e+06  6.974942e+06  576765.0  1725818.0  4384799.0   
2013       51.0  6.200096e+06  7.038774e+06  582684.0  1732621.0  4400477.0   
2014       51.0  6.246342e+06  7.110811e+06  583642.0  1741023.0  4413057.0   
2015       51.0  6.292091e+06  7.181603e+06  586555.0  1746940.5  4424611.0   
2016       51.0  6.335834e+06  7.243849e+06  585501.0  1757121.0  4436974.0   

                             
            75%         max  
year                         
2010  6654375.0  37332685.0  
2011  6717221.5  37676861.0  
2012  6776617.0  38011074.0  
2013  6837396.0  38335203.0  
2014  6902053.5  38680810.0  
2015  6988927.5  38993940.0  
2016  7109535.5  39250017.0

In [59]:
#group data by region, division, and year and summarize it
reshaped_census.groupby(['region','division','year']).describe()

population                            \
                                     count        mean           std   
region  division           year                                        
Midwest East North Central 2010        5.0   9288169.4  3.115322e+06   
                           2011        5.0   9301433.8  3.109807e+06   
                           2012        5.0   9314559.0  3.104953e+06   
                           2013        5.0   9332093.0  3.099620e+06   
                           2014        5.0   9346265.8  3.091076e+06   
...                                    ...         ...           ...   
West    Pacific            2012        5.0  10185665.0  1.574263e+07   
                           2013        5.0  10274464.0  1.587648e+07   
                           2014        5.0  10371286.2  1.601909e+07   
                           2015        5.0  10468346.0  1.614496e+07   
                           2016        5.0  10560386.6  1.624360e+07   

                                                                              \
                                       min        25%        50%         75%   
region  division           year                                                
Midwest East North Central 2010  5690263.0  6490528.0  9877495.0  11540983.0   
                           2011  5709640.0  6516480.0  9876213.0  11544824.0   
                           2012  5726177.0  6537743.0  9887238.0  11550839.0   
                           2013  5742854.0  6569102.0  9898982.0  11570022.0   
                           2014  5758377.0  6595233.0  9915767.0  11594408.0   
...                                    ...        ...        ...         ...   
West    Pacific            2012   731089.0  1391820.0  3899116.0   6895226.0   
                           2013   736879.0  1406481.0  3925751.0   6968006.0   
                           2014   736705.0  1416349.0  3968371.0   7054196.0   
                           2015   737709.0  1425157.0  4024634.0   7160290.0   
                           2016   741894.0  1428557.0  4093465.0   7288000.0   

                                             
                                        max  
region  division           year              
Midwest East North Central 2010  12841578.0  
                           2011  12860012.0  
                           2012  12870798.0  
                           2013  12879505.0  
                           2014  12867544.0  
...                                     ...  
West    Pacific            2012  38011074.0  
                           2013  38335203.0  
                           2014  38680810.0  
                           2015  38993940.0  
                           2016  39250017.0  

[63 rows x 8 columns]